urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/VersionChecker.java

```diff
@@ -0,0 +1,132 @@
+        {
+            if (!update)
+            {
+                final File file = new File(api.getMinecraftDirectory(),
+                        (String) version.get("file"));
+                final String lastVersion = getLastVersion(api,
+                        (String) version.get("source"));
+                
+                if (lastVersion != null)
+                {
+                    if (!file.exists())
+                    {
+                        update = true;
+                        api.getUpdater().setAskUpdate(false);
+                        api.getUpdater().setDoUpdate(true);
+                    }
+                    else
+                    {
+                        final String current = readVersionFile(api, file);
+                        final Version v1 = Version.parseString(current);
+                        final Version v2 = Version.parseString(lastVersion);
+                        
+                        if (v2.compareTo(v1) > 0)
+                        {
+                            update = true;
+                            boolean force = (Boolean) version.get("force");
+                            if (!force)
+                            {
+                                force = api.getConfig().getBoolean(
+                                        "updater.askUpdateIfAvailable", true);
+                            }
+                            
+                            if (!force)
+                            {
+                                api.getUpdater().setAskUpdate(true);
+                            }
+                            api.getUpdater().setDoUpdate(true);
+                        }
+                    }
+                }
+                
+                if (update)
+                {
+                    updateVersionFile(file, lastVersion);
+                }
+            }
+        }
+    }
+    
+    public static String readVersionFile(LauncherAPI api, File file)
+            throws Exception
+    {
+        if (!file.exists())
+        {
+            file.createNewFile();
+            updateVersionFile(file, "-1");
+        }
+        
+        String version = null;
+        
+        final DataInputStream in = new DataInputStream(
+                new FileInputStream(file));
+        version = in.readUTF();
+        in.close();
+        
+        return version;
+    }
+    
+    public static void updateVersionFile(LauncherAPI api, File file,
+            Map<String, Object> version) throws Exception
+    {
+        updateVersionFile(file,
+                getLastVersion(api, (String) version.get("source")));
+    }
+    
+    public static void updateVersionFile(File file, String lastVersion)
+            throws Exception
+    {
+        if (!file.exists())
+        {
+            file.createNewFile();
+        }
+        
+        final DataOutputStream out = new DataOutputStream(new FileOutputStream(
+                file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource created by the following constructor: *FileOutputStream*. The resource is referenced at lines 107, 109. The resource is closed at line 110. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.writeUTF throws an exception. Ensure the object returned by *FileOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).